src/core/derivation.ts

```diff
@@ -1,20 +1,25 @@
-import {IObservable, IDepTreeNode, propagateReadiness, propagateStaleness, addObserver, removeObserver} from "./observable";
+import {IObservable, IDepTreeNode, addObserver, removeObserver} from "./observable";
 import {globalState, resetGlobalState} from "./globalstate";
 import {invariant} from "../utils/utils";
 import {isSpyEnabled, spyReport} from "./spy";
-import {SimpleSet, ISetEntry} from "../utils/set";
+import {ISetEntry} from "../utils/set";
+import {ComputedValue} from "./computedvalue";
 
 /**
  * A derivation is everything that can be derived from the state (all the atoms) in a pure manner.
  * See https://medium.com/@mweststrate/becoming-fully-reactive-an-in-depth-explanation-of-mobservable-55995262a254#.xvbh6qd74
  */
-export interface IDerivation extends IDepTreeNode, IObservable, ISetEntry {
+export interface IDerivation extends IDepTreeNode, ISetEntry {
 	observing: IObservable[];
-	staleObservers: IDerivation[];
-	observers: SimpleSet<IDerivation>;
-	dependencyStaleCount: number;
-	dependencyChangeCount: number;
-	onDependenciesReady(): boolean;
+	newObserving: IObservable[];
+	/**
+	 * Describes state of observing dependencies to know whet it's needed to rerun
+	 * -1 <- not tracking dependencies
```

Comment:
using an enum would make this cleared: `enum IDerivationState { NOT_TRACKING = -1, UP_TO_DATE = 0, POSSIBLY_STALE = 1, STALE = 2 }`